### ISSUE TITLE
Not all props are logged to console

### ISSUE THREAD

**Issue by sandervspl** [CompuIves/codesandbox-client#408]
I tried to log my props to the client's console, and I noticed the object is not complete. The browser's console does however show the complete props object.

Example sandbox: https://codesandbox.io/s/7wymwmq80j

Not sure if other objects are affected.

**Comment by CompuIves** [comment]
Thanks for reporting this! Seems like we should check out our JSON.stringify here: https://github.com/CompuIves/codesandbox-client/blob/master/packages/app/src/sandbox/console/transform-json.js#L21 to see why several properties are not stringified.

**Comment by lukePeavey** [comment]
Is it ok if i take a look at this?

**Comment by CompuIves** [comment]
Yep, absolutely @lukePeavey!

**Comment by lukePeavey** [comment]
Ill do some work on it today and see if i can figure it out.

On Dec 21, 2017 7:37 AM, "Ives van Hoorne" <notifications@github.com> wrote:

Yep, absolutely @lukePeavey <https://github.com/lukepeavey>!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/CompuIves/codesandbox-client/issues/408#issuecomment-353340531>,
or mute the thread

**Comment by lbogdan** [comment]
That's simply because `JSON.stringify()` ignores functions.

**Comment by xyc** [comment]
Is it necessary to stringify/parse messages?

**Comment by lbogdan** [comment]
@xyc Yes, because the sandbox preview is loaded in an iframe, running in a different JS context than the console, and you can't just pass objects across contexts.

One solution would be to **compose** the console message data directly in the sandbox preview, including functions and other non-passable data, **stringify it and send it** to the console that would just **display** it (right now it's the console that both **composes and displays** the messages).

**Comment by CompuIves** [comment]
> One solution would be to compose the console message data directly in the sandbox preview, including functions and other non-passable data, stringify it and send it to the console that would just display it (right now it's the console that both composes and displays the messages).

Interesting, what do you mean exactly with `compose`? One approach I was thinking of was traversing all keys and stringify it based on type (eg. for `function` we'd do `toString`). Maybe yours is better

**Comment by lbogdan** [comment]
I wasn't thinking about a specific approach, but yeah, something along the lines of correctly serializing the console message **before** in the sandbox preview, before passing it to the console, so pretty what you said. :smile:

**Comment by xyc** [comment]
@lbogdan Got it, so a reliable way to serialize/deserialize functions is needed and then it can be passed betwen sandbox frame and console.

**Comment by CompuIves** [comment]
Exactly @xyc!

**Comment by xyc** [comment]
@CompuIves Maybe an alternative approach is to make react-inspector fetch its data async and "query" the sandbox frame as needed. Just an idea.

**Comment by CompuIves** [comment]
Hmm, that makes a lot more sense @xyc, a much more sophisticated approach. It would also solve the circular structure problems. I think we should go this way instead, more reliable.
